app/controllers/api/storage.php

```diff
@@ -1154,49 +1371,98 @@
 
+            if($end == null) {
+                $end =  min(($start + 2000000-1), ($size - 1));
+            }
+
+            if($unit != 'bytes' || $start >= $end || $end >= $size) {
+                throw new Exception('Invalid range', 416);
+            }
+
+            $response
+                ->addHeader('Accept-Ranges', 'bytes')
+                ->addHeader('Content-Range', 'bytes ' . $start . '-' . $end . '/' . $size)
+                ->addHeader('Content-Length', $end - $start + 1)
+                ->setStatusCode(Response::STATUS_CODE_PARTIALCONTENT);
+        }
+
+        $source = '';
         if (!empty($file->getAttribute('openSSLCipher'))) { // Decrypt
+            $source = $device->read($path);
             $source = OpenSSL::decrypt(
                 $source,
                 $file->getAttribute('openSSLCipher'),
-                App::getEnv('_APP_OPENSSL_KEY_V'.$file->getAttribute('openSSLVersion')),
+                App::getEnv('_APP_OPENSSL_KEY_V' . $file->getAttribute('openSSLVersion')),
                 0,
                 \hex2bin($file->getAttribute('openSSLIV')),
                 \hex2bin($file->getAttribute('openSSLTag'))
             );
         }
 
-        $output = $compressor->decompress($source);
-        $fileName = $file->getAttribute('name', '');
+        if (!empty($file->getAttribute('algorithm', ''))) {
+            if(empty($source)) {
+                $source = $device->read($path);
+            }
+            $compressor = new GZIP();
+            $source = $compressor->decompress($source);
+        }
 
         $usage
             ->setParam('storage.files.read', 1)
             ->setParam('bucketId', $bucketId)
         ;
 
-        $response
-            ->setContentType($contentType)
-            ->addHeader('Content-Security-Policy', 'script-src none;')
-            ->addHeader('X-Content-Type-Options', 'nosniff')
-            ->addHeader('Content-Disposition', 'inline; filename="'.$fileName.'"')
-            ->addHeader('Expires', \date('D, d M Y H:i:s', \time() + (60 * 60 * 24 * 45)).' GMT') // 45 days cache
-            ->addHeader('X-Peak', \memory_get_peak_usage())
-            ->send($output)
-        ;
+        if(!empty($source)) {
+            if(!empty($rangeHeader)) {
+                $response->send(substr($source, $start, ($end - $start + 1)));
+            }
+            $response->send($source);
+        }
+
+        if(!empty($rangeHeader)) {
+            $response->send($device->read($path, $start, ($end - $start + 1)));
+        }
+
+        $size = $device->getFileSize($path);
+        if ($size > APP_STORAGE_READ_BUFFER) {          
+            $response->addHeader('Content-Length', $device->getFileSize($path));
+            $chunk = 2000000; // Max chunk of 2 mb
+            for ($i=0; $i < ceil($size / $chunk); $i++) {
+                $response->chunk($device->read($path, ($i * $chunk), min($chunk, $size - ($i * $chunk))), (($i + 1) * $chunk) >= $size);
```

Comment:
No way I am reviewing lines like this 😅 Can this be hidden behind some utopia methods for better readability of the code? @eldadfux WDYT?

Reply:
Don't think we can fit this into any library, as we are already using two libraries here,
1. Storage to read a chunk
2. swoole/framework to send a chunk in response
We need both, and don't want to pass around one library object to another.